library/Icinga/Web/RememberMe.php

```diff
@@ -0,0 +1,188 @@
+            $newData[$key] = $value;
+        }
+
+        $rememberMe->rsa = (new RSA())->loadKey($newData['private_key'], $publicKey);
+        $rememberMe->username = $rememberMe->rsa->decryptFromBase64($data[0]);
+        $rememberMe->encryptedPassword = $data[1];
+
+        return $rememberMe;
+    }
+
+    /**
+     * Encrypt the given username, password and assign the variables
+     *
+     * @param $username
+     * @param $password
+     *
+     * @return static
+     */
+    public static function fromCredentials($username, $password)
+    {
+
+        $rememberMe = new static();
+
+        $rsa = (new RSA())->loadKey(...RSA::keygen());
+
+        $rememberMe->encryptedPassword = $rsa->encryptToBase64($password);
+        $rememberMe->username = $username;
+        $rememberMe->rsa = $rsa;
+
+        return $rememberMe;
+    }
+
+    /**
+     * Set the values for 'remember-me' cookie
+     *
+     * @return Cookie
+     */
+    public function getCookie()
+    {
+        $value = $this->rsa->encryptToBase64($this->username);
+        $value[] = $this->encryptedPassword[0];
+        $value[] = base64_encode($this->rsa->getPublicKey());
+
+        return (new Cookie(static::COOKIE))
+            ->setExpire(time() + 60 * 60 * 24 * 30)
+            ->setHttpOnly(true)
+            ->setValue(implode('|', $value));
+    }
+
+    /**
+     * Unset the values for 'remember-me' cookie
+     *
+     * @return Cookie
+     */
+    public static function unsetCookie()
+    {
+        unset($_COOKIE[static::COOKIE]);
+
+        return (new Cookie(static::COOKIE))
+            ->setExpire(1)
+            ->setHttpOnly(true)
+            ->setValue(null);
+    }
+
+    /**
+     * Authenticate the given username and password
+     *
+     * @return bool     True if authentication succeed, false if not
+     *
+     * @throws \Icinga\Exception\AuthenticationException
+     */
+
+    public function authenticate()
+    {
+        list($password) = $this->rsa->decryptFromBase64($this->encryptedPassword);
+        $auth = Auth::getInstance();
+        $authChain = $auth->getAuthChain();
+        $authChain->setSkipExternalBackends(true);
+        $user = new User($this->username[0]);
+        if (! $user->hasDomain()) {
+            $user->setDomain(Config::app()->get('authentication', 'default_domain'));
+        }
+        $authenticated = $authChain->authenticate($user, $password);
+        if ($authenticated) {
+            $auth->setAuthenticated($user);
+        }
+
+        return $authenticated;
+    }
+
+    /**
+     * Database insert for the given private and public key
+     *
+     * Save new keys in database
+     */
+    public function persist()
+    {
+        $this->getDb()->insert('rememberme', [
+            'username' => $this->username,
+            'private_key' => $this->rsa->getPrivateKey(),
+            'public_key' => $this->rsa->getPublicKey(),
+            'ctime' => date('Y-m-d H:i:s'),
```

Comment:
Let the database handle the timestamp with `new Expression('NOW()')`. Note that `Expression` is `ipl/Sql/Expression`. The same goes for `mtime`.